src/jit/decomposelongs.cpp

```diff
@@ -946,6 +949,72 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
 }
 
 //------------------------------------------------------------------------
+// DecomposeMul: Decompose GT_MUL
+//
+// Arguments:
+//    use - the LIR::Use object for the def that needs to be decomposed.
+//
+// Return Value:
+//    The next node to process.
+//
+GenTree* DecomposeLongs::DecomposeMul(LIR::Use& use)
+{
+    assert(use.IsInitialized());
+
+    GenTree*   tree = use.Def();
+    genTreeOps oper = tree->OperGet();
+
+    assert(oper == GT_MUL);
+    assert((tree->gtFlags & GTF_MUL_64RSLT) != 0);
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+
+    GenTree* loOp1 = op1->gtGetOp1();
+    GenTree* hiOp1 = op1->gtGetOp2();
+    GenTree* loOp2 = op2->gtGetOp1();
+    GenTree* hiOp2 = op2->gtGetOp2();
+
+    Range().Remove(hiOp1);
+    Range().Remove(hiOp2);
+    Range().Remove(op1);
+    Range().Remove(op2);
+
+    // Get rid of the hi ops. We don't need them.
+    tree->gtOp.gtOp1 = loOp1;
+    tree->gtOp.gtOp2 = loOp2;
+    tree->gtOper = GetHiOper(oper);
+
+    if (use.IsDummyUse())
+        return use.Def()->gtNext;
```

Comment:
It looks like everything from here down is copy-and-paste from DecomposeCall and should be refactored so the two share the code.